### ISSUE TITLE
Use WebEngine instead of WebKit for widget applications

### ISSUE THREAD

**Issue by letsfindaway** [letsfindaway/OpenBoard#14]
I'm using this issue to document my considerations and approach for the switch from `QWebKit` to `QWebEngine` for widget applications. The feature is developed at https://github.com/letsfindaway/OpenBoard/tree/refactor-webengine.

WebKit is no longer developed and is deprecated as Qt API, so a move is necessary in the middle term. For porting, see https://doc.qt.io/qt-5/qtwebenginewidgets-qtwebkitportingguide.html

Useful information also here: https://doc.qt.io/qt-5/qtwebengine-features.html.

For the refactoring I'm using a "depth first" approach, i.e. I'm first refactoring the `UBGraphicsWidgetItem` as far as possible and try to understand and solve most of the problems and incompatibilities occurring with the switch of the web engine. Then in a next step I will address `UBToolWidget` and finally the embedded web browser, considering the preexisting work on this.

## Add dependency in project file and build scripts

As the first step is only for the widget applications and not for the use of WebKit as browser, I will add a new dependency in the project file:

```
QT += webenginewidgets
```

As I'm not familiar with the build scripts I leave it to othes to add this dependency there. However I will add it to my project in the openSUSE build service once it is in a stage to be deployed there.

- [x] Add `webenginewidgets` to project file
- [x] Add WebEngine to build dependencies in scripts and packages (done later before release, not in this PR)

## Analysis of usage of QWebKit

The following classes are used from QWebKit for widget applications:

- `QWebView` to be replaced with `QWebEngineView`
- `QWebFrame` to be replaced by `QWebEngineView`, which now covers this, too
- `QWebPage` to be replaced by `QWebEnginePage`
- `QWebSettings` to be replaced by `QWebEngineSettings`

Additionally from the `src/web` directory and currently not covered here:

- `QWebPluginFactory`
- `QWebElement`
- `QWebElementCollection`
- `QWebHitTestResult`
- `QWebHistory`
- `QWebHistoryInterface`

## Use a QGraphicsProxyWidget to wrap the QWebEngineView

The support for widget applications is implemented in `UBGraphicsWidgetItem`, which is a subclass of `QGraphicsItem`. As the web engine does not support a `QGraphicsItem` we have to wrap the `QWebEngineView` in a `QGraphicsProxyWidget`. The `UBGraphicsWidgetItem` will inherit from `QGraphicsProxyWidget` and the `QWebEngineView` will be a member of that class, which is forwarding all necessary calls to that member. After attaching a widget to the proxy, both share ownership. So when the proxy is destroyed, the widget is also automatically destroyed.

However some mechanisms are not supported by the WebEngine. We have to look for alternatives.

- [x] Wrap `QWebEngineView` in `QGraphicsProxyWidget`
- [x] Also use WebEngine in `UBToolWidget`
- [x] Look for alternative for `QPalette` setting: `QWebEnginePage` has a `backgroundColor` property.
- [x] Look for alternative for `OfflineWebApplicationCacheEnabled` (enabled by default for QWebEngine)
- [x] Look for alternative for `OfflineStorageDatabaseEnabled` (enabled by default for QWebEngine)
- [x] Look for alternative for `DeveloperExtrasEnabled`
- [x] Look for alternative for `setNetworkAccessManager` (only necessary if we allow an app widget to access an external URL which requires authentication. Is this realistic?)
- [x] Look for alternative for `setAcceptHoverEvents` -> should be invoked at the `QGraphicsItem`

According to my first tests, it is **_not_** necessary to request a repaint through a timer. It is sufficient to invoke `update()` once loading of the widget is finished. Further updates work flawlessly.

## Injecting the API

The WebKit signal `javaScriptWindowObjectCleared` to inject the JavaScript API at the beginning of the loading process is not supported. Instead WebEngine offers other ways to handle this. It may be worth to consider using a `QWebEngineProfile` shared between all pages. Also all settings will then move to this profile. Using the `scripts()` method gives access to a `QWebEngineScriptCollection`, where the API can be injected once and then used by all pages using this profile. The profile is specified when creating the page.

WebKit is using `QWebFrame::addToJavaScriptWindowObject` to inject a `QObject` properties and slots as JavaScript object. WebEngine is using a `QWebChannel` for this purpose. See e.g. https://stackoverflow.com/questions/28565254/how-to-use-qt-webengine-and-qwebchannel#36601118 and https://stackoverflow.com/questions/31928444/qt-qwebenginepagesetwebchannel-transport-object.

Currently the API objects are attached to the WebView on `javaScriptWindowObjectCleared`. With `QWebChannel` however the API objects can be attached to the channel beforehand, even before the page is loaded.

On the client side however it is necessary to add and instantiate the channel API by adding something like

```html
<script type="text/javascript" src="qrc:///qtwebchannel/qwebchannel.js"></script>

<script type="text/javascript">
    <!-- it's a good idea to initialize webchannel after DOM ready, if your code is going to manipulate the DOM -->
    document.addEventListener("DOMContentLoaded", function () {
        new QWebChannel(qt.webChannelTransport, function (channel) {
            var jshelper = channel.objects.jshelper;
            // do what you gotta do
        });
    });
</script>
```

Instead of adding this code in each widget, we can also inject it to the page. The `UBGraphicsWidgetItem` injects the content of `qwebchannel.js` and from a file `etc/asyncAPI.js` to each widget. This code is executed before any other JavaScript code on the page.

Within `etc/asyncAPI.js` we create the channel and expose the channel objects `widget` and/or `sankore` globally. After that, all property accesses and all function invocations without return value will work. Functions returning a value are handled differently, see also below.

- [x] Find out how to use a `QWebChannel` to inject a `QObject`. Note: channels are per page and must be attached to the page after creating it.
- [x] Add the API objects to the channel in an earlier stage, i.e. before `load()` is called. They must however be added after the item is attached to a scene. We do it now with the `loadStarted` signal.
- [x] Inject the proper JavaScript to make the channel objects available for the widget applications in the same way as `QWebFrame::addToJavaScriptWindowObject` offered it.
- [x] Determine when and where the `QWebEngineProfile` is created.
- [x] Add all settings to the profile
- [x] Add the API scripts to the profile
- [x] Create pages using the profile
- [x] Make all API property access functions `const` and the properties `CONSTANT` to avoid warnings from the `QWebChannel`
- [x] Slot API calls with return type are problematic. The returned value cannot be received synchronously, but only in a callback function added as additional parameter. This makes the invocation syntax used in many application widgets unusabe. See the discussion and proposal below.

## API functions

The API exposed over `QWebChannel` only supports asynchronous operation. It enables

- signals, which are emitted from the C++ side and trigger callbacks at the JS side.
- slots, which are emitted at the JS side and invoke functions at the C++ side.
- slots returning value, which is delivered asynchronously.
- properties, which are cached at the JS side, i.e. when a property changes at the C++ side, it is immediately forwarded over the `QWebChannel` to the JS side. Reading the property just means reading from the cache.

It does however not support functions with parameters and a return value, as the current API uses e.g. for the `QString property(QString key, QString default)` function. Such functions will return their result asynchronously in a callback, which can be supplied as additional parameter to the function call. To make usage even easier, the `etc/asyncAPI.js` script can contain wrapper functions, which wrap the callback with a `Promise`. This enables us to use the `async`/`await` pattern as explained below.

The following is a list of API calls returning a value, which are not properties (may be incomplete):

- sankore.pageThumbnail
- sankore.preference
- sankore.preferenceKeys
- sankore.language
- sankore.isDropableData
- sankore.documentDatastoreApi.key
- sankore.documentDatastoreApi.getItem
- widget.w3cwebstorageApi.key
- widget.w3cwebstorageApi.getItem
- widget.preferenceApi.key
- widget.preferenceApi.getItem

However I do not see any use of these functions besides `preference` and `preferenceKeys`. The interactions also use `language()`, which can however easily be replaced with the property `lang` containing the same value. Here an example, taken from StopWatch, how the new API functions are used:

```javascript
async function init(){
	
    var tempHours;
    var tempMinutes;
    var tempSeconds;
	
    if(window.sankore.async){  // new async API
        tempHours = await window.sankore.async.preference("hours","00");
        tempMinutes = await window.sankore.async.preference("minutes","00");
        tempSeconds = await window.sankore.async.preference("seconds","00");
    } else {  // no API at all
        tempHours = "00";
        tempMinutes = "00";
        tempSeconds = "00";
    }
```

- The function itself needs the `async` keyword.
- Invocations have to use `await` and the `window.sankore.async` functions.

- [x] Which API calls are still relevant? What is obsolete?

## Web Inspector

The QWebEngine comes with a full-fledged web inspector as all Chromium based browsers. This can be very handy to debug application widgets. The following changes enable the web inspector for widgets:

- Reactivate the `GF_SHOW_CONTENT_SOURCE` flag, which currently seems to have no use. We set this flag for a `UBGraphicsWidgetItemDelegate` if debugging is enabled via a configuration flag. This adds a widget menu entry for this action.
- Overwrite `gotoContentSource`. In this function we
  - create a new `QMainWindow` with a `QWebEngineView` as central widget.
  - set the devToolsPage of the application widget page to the page of the new `QWebEngine`.
  - trigger the `inspectElement` action.

We then get a new top-level window containing the web inspector for the widget. The window is closed when clicking to the close button in the title bar or when the associated widget is removed from the board. Selecting the widget menu item a second time brings an existing inspector window to the foreground.

**Comment by letsfindaway** [comment]
Open questions:

- Which role has `UBToolWidget`? What exactly is meant by making a tool from a application widget?

**Comment by letsfindaway** [comment]
# Questions to package maintainers

@kaamui: For some of the questions I have I need the input of the package maintainers. I will collect them in this comment to my issue.

## Widget API

As explained above, the way to communicate between the Qt application and the JS widgets has changed from a synchronous approach with QWebKit to an asynchronous approach with QWebEngine. This makes it impossible to invoke API functions with synchronous return values. Instead a callback is invoked when the result is available. This requires changes in the widget code. How shall this be handled? I can update all the applications accordingly, but

- [ ] How are the original authors of the widgets involved in this process?
- [ ] It is at all acceptable to change the API usage?
- [ ] What parts of the API are obsolete? For some stuff I did not see any usage.

## Obsolete code

Of course I do not want to spend much effort porting obsolete code. So I'm interested whether the following features are still used. See also above.

- [ ] Are those "Apple Widgets" still used? I did not find any use of this. Seems to be related to a file extension of .wdgt. Can it be removed?
- [ ] Is the NPAPI wrapper code still used? Seems to be used to convert some old plugins, but this might be obsolete now.
- [ ] Is `UBGraphicsProxyItem` obsolete? There is no active use of it allover the code and I would like to remove it.

## Browser features

Normally widgets are quite restricted and just work within their assigned area. If they however contain external links, this can be different.

- [ ] What should happen when a link is clicked in the Web Browser widget application which opens a new window or downloads a file? 
- [ ] For the built-in browser: Is there already a preference or even decision whether to keep it or remove it in the middle term?

**Comment by letsfindaway** [comment]
# Next steps

- [ ] Finish the work on the API injection. Widget porting to the async methods can start then.
- [ ] Remove `UBGraphicProxyWidget`. Check and adapt all remaining uses of that and of type checks for `QGraphicsProxyWidget`.
- [ ] Port `UBToolWidget`. Widgets should then work also in the tool mode.
- [ ] Analyse and step-by-step integrate the porting of the web browser as already done by @JBBgameich. 
- [ ] Improve interaction of widgets and browser ("open in new tab" for links in widgets opens browser tab).

**Comment by JBBgameich** [comment]
I assumed that TrapFlash is about flash player widgets and removed them in my QtWebEngine prototype. If it's really just about flash player I still think that's safe to do. Does the feature also work with other kinds of widgets?

**Comment by letsfindaway** [comment]
> I assumed that TrapFlash is about flash player widgets and removed them in my QtWebEngine prototype. If it's really just about flash player I still think that's safe to do. Does the feature also work with other kinds of widgets?

Yes, it does. The name is really misleading. If you check the tool in the current OpenBoard, e.g. by opening any page in the embedded web browser and then clicking the top-most icon in the web browser toolbar, then you see that you can select the whole page or any item in an `<embed>` tag on that page.

This `<embed>` tag is deprecated since HTML 4, but was used for Flash content. Therefore the name. Now the "trap flash" will more or less just able to select the "whole page". This page will then be wrapped into an application widget and placed on the current scene. You can still use the web page there, click on links or whatever, which seems to be quite useful. So I definitely want to keep that feature.

What I did until now and plan to do: I created a complete diff of your webengine branch against the branching point. I will go through the files, probably starting from the core web browser classes, try to understand them, see what you did, and then apply your changes and add mine. On that way I will try to remove as less as possible. Let's see how this will work out!

**Comment by JBBgameich** [comment]
Thank you for all the work you are putting into this! My branch is unfortunately quite messy right now as it was really just done to see how much needs to be done for porting to QtWebEngine. The approach was basically remove the webkit dependency and fix / comment out things until it builds :)
So your approach is definitely right, the changes need a careful review.

**Comment by letsfindaway** [comment]
# Redesign of the UBOEmbedParser

The `UBOEmbedParser` is currently responsible for the following things:

* Parsing the HTML of a web page for `<link>` tags pointing to oembed information.
* Retrieving this information.
* Decoding it as XML of JSON.

It is currently attached as a single instance to the `UBWebController`, which makes switching of tabs and keeping the parse results difficult.

I will redesign it as follows:

- [ ] Attach an instance of the `UBOEmbedParser` to each tab of the browser as soon as the tab is created. The `TabView` will be the parent of the `UBOEmbedParser`, so the parser is automatically destroyed when the tab is closed.
- [ ] Move the extraction of the HTML of a page from `UBWebController` to the `UBOEmbedParser`.
- [ ] Connect the parser to the `loadFinished` signal of the tab's `QWebEngineView`.
- [ ] Keep all parsing results in the `UBOEmbedParser` and make them accessible.
- [ ] Emit signal when parsing is complete.

The role of the `UBWebController` will be:

- [ ] Create the `UBOEmbedParser` when a new tab is created.
- [ ] Keep the enabled state of the `actionWebOEmbed` in sync with the parse results of the current tab.
- [ ] Handle the `actionWebOEmbed`.
